### PR TITLE
add allow_head to add_get

### PR DESCRIPTION
**Pull request by samuelcolvin**
## What do these changes do?

Add the `allow_head` keyword argument for `add_get` to encourage aiohttp server applications to correctly implement http and allow `HEAD` requests to endpoints generally used for `GET` requests. 

Also adds `app.router.set_defaults(allow_head=True)` to switch behaviour of all endpoints for the app.

## Related issue number

#1618

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] Add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new entry to `CHANGES.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using `#issue_number` format at the end of changelog message. Use Pull Request number if there are no issues for PR or PR covers the issue only partially.

### CODE REVIEW

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -755,6 +756,9 @@ def routes(self):
     def named_resources(self):
         return MappingProxyType(self._named_resources)
 
+    def set_defaults(self, *, allow_head):
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: I doubt if we need default for the flag.
It might lead to many confusing cases, e.g. if third party library has something like `register_routes(router, prefix)` method its behavior varies depending on outer flag.

Comment (samuelcolvin) [reply aiohttp/web_urldispatcher.py]: Humm, ok I'll remove it if you prefer. But the current setup is wrong according to the RFC and without the option to set the default you have to explicitly set `allow_head=True` on every route you add:

[RFC 7231](https://tools.ietf.org/html/rfc7231#page-25):

> The server SHOULD send the same
   header fields in response to a HEAD request as it would have sent if
   the request had been a GET, except that the payload header fields
   (Section 3.3) MAY be omitted.

For me, the simplest, clearest and most correct option would be to have `allow_head=True` by default.

**Comment by samuelcolvin** [comment]
~~This is failing due to an error in docs I've fixed in #1669~~ fixed

File: aiohttp/web_urldispatcher.py (hunk 1 above)
Comment (fafhrd91) [reply aiohttp/web_urldispatcher.py]: problem is, we currently do not control when handler writes body, so in most cases
developer needs to support `head` explicitly

Comment (samuelcolvin) [reply aiohttp/web_urldispatcher.py]: The developer can choose to support head explicitly but nothing breaks if they don't.

Allowing HEAD to endpoints which act like they got a GET request and still write body doesn't break anything.

Both django and flask allow users to still (by default) write body even for HEAD requests. I'm not saying they're always right, but their solution is more correct than aiohttp currently.

Here's an example:

**app.py:**
```python
import asyncio
from aiohttp import web

async def handle(request):
    print('preparing slow response...')
    await asyncio.sleep(2, loop=request.app.loop)
    return web.Response(body=b'this is the body\n')

app = web.Application()
app.router.add_get('/', handle, allow_head=True)

web.run_app(app, host='localhost', port=8000)
```

**shell:**
```
~ ➤  curl -v http://localhost:8000
* Rebuilt URL to: http://localhost:8000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET / HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Length: 17
< Content-Type: application/octet-stream
< Date: Wed, 22 Feb 2017 10:08:03 GMT
< Server: Python/3.5 aiohttp/2.0.0a0
< 
this is the body
* Connection #0 to host localhost left intact
~ ➤  curl -I http://localhost:8000
HTTP/1.1 200 OK
Content-Length: 17
Content-Type: application/octet-stream
Date: Wed, 22 Feb 2017 10:08:10 GMT
Server: Python/3.5 aiohttp/2.0.0a0

~ ➤  
```

Nothing went wrong here. 

The app could choose to notice the HEAD request and not spend a lot of time/resources writing body, but I don't see any problem if those not concerned by this (fairly niche) optimisation don't bother with it.

**In summary:** your current approach explicitly contravenes the HTTP RFC. The issue you're worried about is an optimisation which will apply in a very small percentage of requests and where app authors already have all the tools they need to achieve the optimization.

Comment (fafhrd91) [reply aiohttp/web_urldispatcher.py]: Ok. Let's remove url dispatch part and change default allow_head to True

Comment (samuelcolvin) [reply aiohttp/web_urldispatcher.py]: Great. I've updated the PR.

**Comment by fafhrd91** [comment]
very good
